primitives/src/pcs/multilinear_kzg/mod.rs

```diff
@@ -0,0 +1,488 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! Main module for multilinear KZG commitment scheme
+
+mod batching;
+pub(crate) mod srs;
+pub(crate) mod util;
+
+use crate::pcs::{
+    prelude::{
+        Commitment, UnivariateProverParam, UnivariateUniversalParams, UnivariateVerifierParam,
+    },
+    univariate_kzg::KZGUnivariateOpening,
+    PCSError, PolynomialCommitmentScheme, StructuredReferenceString,
+};
+use ark_ec::{
+    msm::{FixedBaseMSM, VariableBaseMSM},
+    AffineCurve, PairingEngine, ProjectiveCurve,
+};
+use ark_ff::PrimeField;
+use ark_poly::{DenseMultilinearExtension, MultilinearExtension};
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize, Read, SerializationError, Write};
+use ark_std::{
+    end_timer, format, marker::PhantomData, rand::RngCore, rc::Rc, start_timer, string::ToString,
+    vec, vec::Vec, One, Zero,
+};
+use batching::{batch_verify_internal, multi_open_internal};
+use srs::{MultilinearProverParam, MultilinearUniversalParams, MultilinearVerifierParam};
+use util::merge_polynomials;
+
+/// KZG Polynomial Commitment Scheme on multilinear polynomials.
+pub struct KZGMultilinearPCS<E: PairingEngine> {
+    #[doc(hidden)]
+    phantom: PhantomData<E>,
+}
+
+#[derive(CanonicalSerialize, CanonicalDeserialize, Clone, Debug)]
+/// proof of opening
+pub struct Proof<E: PairingEngine> {
+    /// Evaluation of quotients
+    pub proofs: Vec<E::G1Affine>,
+}
+
+#[derive(CanonicalSerialize, CanonicalDeserialize, Clone, Debug)]
+/// proof of batch opening
+pub struct BatchProof<E: PairingEngine> {
+    /// The actual proof
+    pub proof: Proof<E>,
+    /// Commitment to q(x):= w(l(x)) where
+    /// - `w` is the merged MLE
+    /// - `l` is the list of univariate polys that goes through all points
+    pub q_x_commit: Commitment<E>,
+    /// openings of q(x) at 1, omega, ..., and r
+    pub q_x_opens: Vec<KZGUnivariateOpening<E>>,
+}
+
+impl<E: PairingEngine> PolynomialCommitmentScheme<E> for KZGMultilinearPCS<E> {
+    // Parameters
+    type ProverParam = (
+        MultilinearProverParam<E>,
+        UnivariateProverParam<E::G1Affine>,
+    );
```

Comment:
Q: is `powers_of_g` in `MultilinearProverParam` and in `UnivariateProverParam` the same thus sharable? or are they used for different purposes?

(maybe I should re-review these implementations when I digest multilinear kzg scheme later.)

Reply:
A: (from Binyi) they can be the same, but conceptually different, plus `univeriate_prover_param` will be relatively small